developer-hub/src/reference_docs/safe_defensive_programming.rs

```diff
@@ -1 +1,615 @@
+    bad_collection_retrieval
+)]
+//!
+#![doc = docify::embed!(
+    "./src/reference_docs/safe_defensive_programming.rs",
+    good_collection_retrieval
+)]
+//!
+//!
+//! ### Defensive Traits
+//!
+//! To also aid in debugging and mitigating the above issues, there is a
+//! [`Defensive`](frame::traits::Defensive) trait (and its companions,
+//! [`DefensiveOption`](frame::traits::DefensiveOption),
+//! [`DefensiveResult`](frame::traits::DefensiveResult))that can be used to defensively unwrap
+//! values.  This can be used in place of
+//! an `expect`, and again, only if the developer is sure about the unwrap in the first place.
+//!
+//! The [`Defensive`](frame::traits::Defensive) trait provides a number of functions, all of which
+//! provide an alternative to 'vanilla' Rust functions:
+//!
+//! - [`defensive_unwrap_or()`](frame::traits::Defensive::defensive_unwrap_or)
+//! - [`defensive_ok_or()`](frame::traits::DefensiveOption::defensive_ok_or)
+//!
+//! The primary difference here is when `debug_assertions` are enabled, that they panic, but in
+//! production/release, they will merely log an error via the logging instance the runtime is using
+//! (i.e., `log::error`).
+//!
+//! This is useful for testing, but also allows for a more failsafe way to mitigate panics.
+//!
+//! ## Integer Overflow
+//!
+//! The Rust compiler prevents any sort of static overflow from happening at compile time, for
+//! example:
+//!
+//! ```ignore
+//! let overflow = u8::MAX + 10;
+//! ```
+//! Would yield the following error:
+//! ```text
+//! error: this arithmetic operation will overflow
+//!    --> src/main.rs:121:24
+//!     |
+//! 121 |         let overflow = u8::MAX + 10;
+//!     |                        ^^^^^^^^^^^^ attempt to compute `u8::MAX + 10_u8`, which would overflow
+//!     |
+//! ```
+//!
+//!
+//! However in the runtime context, we don't always have control over what is being supplied as a
+//! parameter. For example, even this simple adding function could present one of two outcomes
+//! depending on whether it is in **release** or **debug** mode:
+#![doc = docify::embed!("./src/reference_docs/safe_defensive_programming.rs", naive_add)]
+//!
+//! If we passed in overflow-able values at runtime, this could actually panic (or wrap, if in
+//! release).
+//! ```ignore
+//! naive_add(250u8, 10u8); // In debug mode, this would panic. In release, this would return 4.
+//! ```
+//!
+//! The Rust compiler would panic in **debug** mode in the event of an integer overflow. In
+//! **release** mode, it resorts to silently _wrapping_ the overflowed amount in a modular fashion,
+//! (hence returning `4`).
+//!
+//! It is actually the _silent_ portion of this behavior that presents a real issue - as it may be
+//! an unintended, but also a very _silent killer_ in terms of producing bugs. In fact, it would
+//! have been better for this type of behavior to produce some sort of error, or even `panic!`, as
```

Comment:
I don't think we should ever suggest a `panic!` should be relatively better behavior here :P 

```suggestion
//! have been better for this type of behavior to produce some sort of error, or even `panic!`, as
```